adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -282,49 +444,38 @@ def _final_callback(self):
             self._update_avg('var_avg', grad_var, theta)
-        self._local_sqr = None
+        self._local_vars.reset()
 
-    def step(self, *args, **kwargs):
-        """
-        Run one optimizer step using Adascale. Essentially just invokes
-        ``optimizer.step(*args, **kwargs)`` with a scaled learning rate.
+    def _calculate_scale_multiplier(self, pg_idx=None) -> float:
+        if pg_idx is None:
+            raise ValueError("Parameter index must be provided.")
+        grad_sqr = float(self._state["norm_avg"][pg_idx])
+        grad_var = float(self._state["var_avg"][pg_idx])
+        gain = (grad_var + grad_sqr) / (grad_var / self._scale + grad_sqr)
+        return gain
 
-        Arguments:
-            args: Positional arguments passed to ``optimizer.step``.
-            kwargs: Keyword arguments passed to ``optimizer.step``.
-        """
-        if (self.is_accumulation_step()):
-            self._current_accumulation_step += 1
-            self._made_step = False
-            return
 
-        self._current_accumulation_step = 0
-        self._made_step = True
-        initial_lr = [pg["lr"] for pg in self._optimizer.param_groups]
-        for idx, param_group in enumerate(self._optimizer.param_groups):
-            grad_sqr = float(self._state["norm_avg"][idx])
-            grad_var = float(self._state["var_avg"][idx])
-            gain = (grad_var + grad_sqr) / (grad_var / self._scale + grad_sqr)
-            param_group["lr"] = gain * param_group["lr"]
-        self._optimizer_step(*args, **kwargs)
-        for lr, param_group in zip(initial_lr, self._optimizer.param_groups):
-            param_group["lr"] = lr
+class LinearScale(ScalingRuleBase):
 
-    def patch_optimizer(self):
-        """
-        Monkey-patch the optimizer's step function with :meth:`AdaScale.step`.
-        """
-        # TODO: detect if the optimizer has already been patched.
+    def _calculate_scale_multiplier(self, pg_index=None) -> float:
+        return self.scale
 
-        @functools.wraps(self._optimizer.step)
-        def wrapper(*args, **kwargs):
-            return self.step(*args, **kwargs)
-        old_zero_grad = self._optimizer.zero_grad
+    def get_progress(self):
+        return self.scale
 
-        @functools.wraps(self._optimizer.zero_grad)
-        def zero_wrapper(*args, **kwargs):
-            if self._made_step:
-                return old_zero_grad(*args, **kwargs)
-            return None
-        self._optimizer.step = wrapper
-        self._optimizer.zero_grad = zero_wrapper
+    def to_tensorboard(self, writer, global_step, tag_prefix=""):
+        writer.add_scalar(tag_prefix + "Learning_Rate_Factor",
+                          self.scale, global_step)
+
+
+class SqrtScale(ScalingRuleBase):
+
+    def _calculate_scale_multiplier(self, pg_index=None) -> float:
+        return math.sqrt(self.scale)
+
+    def get_progress(self):
+        return math.sqrt(self.scale)
```

Comment:
I don't think this is correct mathematically. Progress should be calculated the same way as AdaScale in all cases.